### PR TITLE
tools: build and use setsid command from external source when not available

### LINKED ISSUE #8424: openocd.sh: line 195: setsid: command not found
#### Description

While running openocd for debugging (e.g. make debug) on OS X I have the following:

```console
openocd.sh: line 195: setsid: command not found
```

Not sure if it's an issue but I wonder if the command `setsid` is strictly necessary.

#### Steps to reproduce the issue

Run `make debug` on a fresh installed OS X.
#### Expected results

No errors.
#### Actual results

Error described above.
#### Versions

```console
Installed toolchain versions
----------------------------
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
          native gcc: Apple LLVM version 8.0.0 (clang-800.0.42.1)
          msp430-gcc: msp430-gcc (GCC) 4.7.0 20120322 (mspgcc dev 20120911)
             avr-gcc: avr-gcc (GCC) 7.2.0
   arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2017-q4-major) 7.2.1 20170904 (release) [ARM/embedded-7-branch revision 255204]
    mips-mti-elf-gcc: missing/error
               clang: Apple LLVM version 8.0.0 (clang-800.0.42.1)
arm-none-eabi-newlib: "2.5.0"
 mips-mti-elf-newlib: missing/error
            avr-libc: "2.0.0" ("20150208")
            cppcheck: missing
          coccinelle: missing
                 git: git version 2.16.0
```

BTW the command cited to give the versions doesn't print openocd...

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PRs tries to fix #8424 by introducing some make logic to fetch, build and use setsid from an external repository. Apparently setsid is not available by default on MacOSX.

The Make logic in dist/tools and makefiles should be ok. I'm not sure whether the changes in Makefile.include could be improved.

Maybe the external repository is not the best one but this can be changed easily.

### Testing procedure

1. Start a debug session on Linux (which provides setsid) using the `debug` target. The external setsid should not be used
2. Start a debug session on MacOSX (which doesn't setsid by default). The external setsid should be fetched and built on the fly.

### Issues/PRs references

fixes #8424

### CODE REVIEW

**Comment by aabadie** [comment]
I wrote this PR blindly, since I don't have a Mac but maybe someone with a Mac can test ? @smlng or @kYc0o ?

**Comment by cladmi** [comment]
I think the main problem is "do we even need `setsid`".
But making it work by default on `osx` while we figure that out is a good idea.
Fix the bug first!

There are also other places using `setsid` without using the `SETSID` variables.

File: makefiles/tools/targets.inc.mk [hunk 1]
```diff
@@ -24,3 +24,8 @@ $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb:
 
 mosquitto_rsmb: $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb
 	@make -C $(RIOTTOOLS)/mosquitto_rsmb run
+
+$(RIOTTOOLS)/setsid/setsid:
```
Comment (cladmi) [inline makefiles/tools/targets.inc.mk]: Please also add the dependency to the `Makefile` like for `edbg` to make it rebuild on version change.

File: Makefile.include [hunk 2]
```diff
@@ -466,6 +466,14 @@ define check_cmd
 	    exit 1;}
 endef
 
+# Check setsid command is available
+#   This is not the case on MacOSX, so it must be built on the fly
+ifeq (, $(shell which setsid))
```
Comment (cladmi) [inline Makefile.include]: The `shell which` patterns are a bit different, check the other ones. I think some system provides an error message on error so cannot rely on the output.
```suggestion
ifeq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
```

Comment (aabadie) [reply Makefile.include]: Thanks, but in fact the right line is:
```mk
ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
```

**Comment by aabadie** [comment]
@cladmi, thanks for your feedback. I've updated the PR with your suggestions and replaced remaining occurrences of setsid (with AVR based boards using avarice, which I think could be factorized but well).
It would be great if a Mac user could test this PR.

**Comment by aabadie** [comment]
Maybe @x3ro can also test on Mac ?

**Comment by aabadie** [comment]
@cladmi, do you think this PR is valid to go into the release ? (since it fixes a bug)
If yes, it would be great to have someone testing it.

**Comment by x3ro** [comment]
I can confirm that this worked for me when testing with

     BOARD=stm32f429i-disc1 STLINK_VERSION=2 make debug

**Comment by aabadie** [comment]
@cladmi, your comments are addressed and @x3ro tested it with success.

**Comment by aabadie** [comment]
@cladmi I think your comments are addressed. Maybe @fjmolinas can also confirm that it works (since he works on a mac)

**Comment by Lotterleben** [comment]
@aabadie Just tested on MAC OS, works for me :)

**Comment by cladmi** [comment]
Can you squash ? for a final check of commit messages and all ?

I noticed 3 tools are using the same `debug.sh` file. Could be put in common somehow :D

**Comment by aabadie** [comment]
> Can you squash ? for a final check of commit messages and all ?

finally done.

**Comment by aabadie** [comment]
Another of my low hanging fruit I think. This was tested with success BTW. I can rebase if needed.

**Comment by miri64** [comment]
@smlng can you have a look?

**Comment by aabadie** [comment]
This PR is waiting for an ACK for quite some time now.

**Comment by aabadie** [comment]
Anyone could test and eventually merge this PR ?

File: makefiles/tools/targets.inc.mk [hunk 3]
```diff
@@ -24,3 +24,8 @@ $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb:
 
 mosquitto_rsmb: $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb
 	@make -C $(RIOTTOOLS)/mosquitto_rsmb run
+
+$(RIOTTOOLS)/setsid/setsid: $(RIOTTOOLS)/setsid/Makefile
+	@echo "[INFO] setsid binary not found - building it from source now"
+	@make -C $(RIOTTOOLS)/setsid
+	@echo "[INFO] setsid binary successfully build!"
```
Comment (aabadie) [inline makefiles/tools/targets.inc.mk]: Note to myself: s/build/built/

Comment (aabadie) [reply makefiles/tools/targets.inc.mk]: Just checked and in fact, all other tools have the same typo. Let's leave it this way here and I'll open a follow-up PR to fix them all in a row.

File: Makefile.include [hunk 4]
```diff
@@ -507,6 +507,13 @@ define check_cmd
 	    exit 1;}
 endef
 
+# Check setsid command is available
+#   This is not the case on MacOSX, so it must be built on the fly
+ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
+  export SETSID = $(RIOTTOOLS)/setsid/setsid
```
Comment (fjmolinas) [inline Makefile.include]: @cladmi is this the kind of places where target-export-variables should be used?

File: Makefile.include [hunk 5]
```diff
@@ -507,6 +507,13 @@ define check_cmd
...
+  DEBUGDEPS += $(SETSID)
```
Comment (fjmolinas) [inline Makefile.include]: This could be documented in `vars.inc.mk`

File: makefiles/tools/targets.inc.mk [hunk 6]
```diff
@@ -1,5 +1,5 @@
 # This file contains a collection of targets defined by tools shipped with RIOT.
-# The reason for keeping those targets separated from their tool Makfiles is
+# The reason for keeping those targets separated from their tool Makefiles is
```
Comment (fjmolinas) [inline makefiles/tools/targets.inc.mk]: unrelated

File: makefiles/tools/targets.inc.mk (hunk 3 above)
Comment (fjmolinas) [reply makefiles/tools/targets.inc.mk]: yep

File: Makefile.include (hunk 5 above)
Comment (aabadie) [reply Makefile.include]: Documentation added.

File: makefiles/tools/targets.inc.mk (hunk 6 above)
Comment (aabadie) [reply makefiles/tools/targets.inc.mk]: removed

File: Makefile.include [hunk 7]
```diff
@@ -570,6 +570,15 @@ define check_cmd
 	    exit 1;}
 endef
 
+# Check if setsid command is available on the system for debug target
+# This is not the case on MacOSX, so it must be built on the fly
+ifneq (,$(filter debug, $(MAKECMDGOALS)))
+  ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
+    export SETSID = $(RIOTTOOLS)/setsid/setsid
```
Comment (fjmolinas) [inline Makefile.include]: I would still use `target-export-variables`. Now you are only making the check if `debug` is present in `$(MAKECMDGOALS)` which is good, but if `debug` is present as well as other targets, than it will still be evaluated and SETSID exported for all targets.

If you use `target-export-variables` this will only be exported for `debug`.

Comment (fjmolinas) [reply Makefile.include]: It will translate to:

`debug: export SETSID = $(RIOTTOOLS)/setsid/setsid`

Comment (aabadie) [reply Makefile.include]: using target-export-variables seems to work.

**Comment by aabadie** [comment]
> Feel free to squash after addressing.

Done

**Comment by aabadie** [comment]
Thanks again @fjmolinas !
